### PR TITLE
Make the PPX emit less sections

### PR DESCRIPTION
**Pull request by Drup**
Apparently, the eliom ppx is a bit too eager to put sections everywhere.
Those two small optimisations are backported from the eliom compiler. This is basically untested.

@vouillon, can you take a look?

### CODE REVIEW

**Comment by vasilisp** [comment]
@pwbs could you test if this patch magically resolves #362 ? A bit far-fetched, I know, but the trouble somehow magically arose by the patterns produced by our PPX, and these patterns will now change.

**Comment by Drup** [comment]
@vasilisp I don't think that's far fetched, since that's the reason @vouillon gave me to apply that optim. :)

**Comment by vouillon** [comment]
Indeed, it seems to magically resolves #362 :), at least on the `os.pgocaml` distillery template.

**Comment by vasilisp** [comment]
Great news!

Tthe code looks good.

**Comment by Drup** [comment]
@vouillon I assume you tested on the besport website ? I would prefer some serious testing before we merge that :)

File: src/ppx/ppx_eliom_client.ml [hunk 1]
```diff
@@ -152,9 +163,10 @@ module Pass = struct
       open_client_section loc ::
       register_client_closures client_expr_data @
       define_client_functions loc client_expr_data @
-      [ item ;
-        close_server_section loc ;
-      ]
+      [ item ] @
+      (if flush_injection () then
```
Comment (vouillon) [inline src/ppx/ppx_eliom_client.ml]: The `open_client_section` should be removed. Not the `close_server_section`. Right?

Comment (Drup) [inline src/ppx/ppx_eliom_client.ml]: Indeed! I'll fix that in the afternoon.

**Comment by vouillon** [comment]
So, the second commit is broken.
And I could not reproduce the crash with the os.pgocaml distillery template, so I can't tell after all whether this fixes the problem...
I'm going to try with besport.

File: src/ppx/ppx_eliom_client.ml (hunk 1 above)
Comment (Drup) [inline src/ppx/ppx_eliom_client.ml]: Should now be fixed.

**Comment by vouillon** [comment]
So, the crash still happens with the first commit only.
And the second commit is still broken (this can be reproduced with the os.pgocaml distillery template).

File: src/ppx/ppx_eliom_client.ml [hunk 2]
```diff
@@ -128,7 +138,8 @@ module Pass = struct
   (** Syntax extension *)
 
   let client_str item =
-    if not @@ must_have_section item then [item]
+    if not @@ must_have_section item
+    || flush_injection () then [item]
```
Comment (vouillon) [inline src/ppx/ppx_eliom_client.ml]: That should be `not (must_have_section item && flush_injection ())`

Comment (Drup) [inline src/ppx/ppx_eliom_client.ml]: Oups. Fixed.

**Comment by vouillon** [comment]
The besport website is still crashing on Safari with this change :(

But I think we are still emitting a lot more `close_server_section` than with the camlp4 syntax extension.

**Comment by Drup** [comment]
The thing is, if you have this piece of code:

```ocaml
{client{
let x = 2
let y = 3 
}}
```
the camlp4 extension will consider that as only one section. The ppx consider it has two sections. The reason is to handle that kind of cases correctly:

```ocaml
{shared{
let x = 3 
let y = 3 + {client{ ~%x }}
}}
```
This doesn't work in the camlp4 extension, you must split in two different sections. In the ppx, it'll work directly, since sections are naturally split.

We can probably find a better scheme by analyzing things less locally. I'll try to think about it.

**Comment by Drup** [comment]
Several new ideas:

1. If the declaration doesn't lead to any evaluation (in particular, is a function declaration), we can remove the call to `close`.
2. Successive server sections do not need to be separated by a `closer_server_section`. We can have one `close` after the last successive server section.
3. We emit many constant string that are equals. We could share them.

**Comment by Drup** [comment]
I implemented idea 1 and 3.
Using ocsigen-start's `os.pgocaml` template as a benchmark. 

 |close_server_section|open_client_section|register_client_closure|Constant strings
---|---|---|---|---
Current|747|1053|217|6303
31bcbd8|665|963|217|6135
cd8e323|665|47|217|5219
6bd4f58|665|47|217|5219
4b12707|278|47|217|4832
7099dfb|278|47|217|4577

I can probably reduce it further by fine tuning `Cannot_have_fragment`. I'm a bit disappointed by the limited impact of idea 3.

**Comment by Drup** [comment]
The javascript was compiled with ` -jsopt --noinline -jsopt --disable=shortvar -jsopt --pretty`.

**Comment by Drup** [comment]
Further improvements on idea 1 to eliminate module elements. It looks quite good now. :)

 |close_server_section|open_client_section|register_client_closure|Constant strings
---|---|---|---|---
Current|747|1053|217|6303
288a58c|160|47|217|4546

**Comment by Drup** [comment]
@vouillon Running `grep -o local/var/www/<app>/<app>.js -e '"[^"]*"'` on the `os.pgocaml` example is a bit striking, there is a lot of room for improvement. 

In particular, if we run `grep -o local/var/www/<app>/<app>.js -e '"[^"]*"' | sort | uniq -c | sort -nr` on the compressed version, we can see that there is *259* instances of `"number"`...

**Comment by Drup** [comment]
(For the curious, the size impact is quite small, less than 5% without gziping)

**Comment by vouillon** [comment]
The repetition of the string `"number"` is deliberate. More efficient code is produced for the expression `typeof x == "number"` than for `typeof x == y`  when `y` is bound to `"number"`.

**Comment by vouillon** [comment]
The result is impressive. And Safari no longer crashes!

 |open_client_section|close_server_section
---|---|---
camlp4|3200|2627
ppx(master)|3328|2748
ppx_opt|374|484

There is still room for improvement. For file `os_services.eliom` in `ocsigen-start`, I think we only need one `close_server_section` and one `open_client_section`, but we get the following code:
```
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
close_server_section(eliom_compilation_unit_id_fhytq);
open_client_section(eliom_compilation_unit_id_fhytq);
var main_service$0=get_injection(_akR_,_akQ_,_akP_);
open_client_section(eliom_compilation_unit_id_fhytq);
get_injection(_akU_,_akT_,_akS_);
open_client_section(eliom_compilation_unit_id_fhytq);
get_injection(_akX_,_akW_,_akV_);
open_client_section(eliom_compilation_unit_id_fhytq);
get_injection(_ak0_,_akZ_,_akY_);
open_client_section(eliom_compilation_unit_id_fhytq);
get_injection(_ak3_,_ak2_,_ak1_);
open_client_section(eliom_compilation_unit_id_fhytq);
var connect_service=get_injection(_ak6_,_ak5_,_ak4_);
open_client_section(eliom_compilation_unit_id_fhytq);
var disconnect_service=get_injection(_ak9_,_ak8_,_ak7_);
open_client_section(eliom_compilation_unit_id_fhytq);
var action_link_service=get_injection(_ala_,_ak$_,_ak__);
open_client_section(eliom_compilation_unit_id_fhytq);
var set_password_service=get_injection(_ald_,_alc_,_alb_);
open_client_section(eliom_compilation_unit_id_fhytq);
get_injection(_alg_,_alf_,_ale_);
```

**Comment by Drup** [comment]
@vouillon Nice! You can merge after you reviewed/when you feel like it, so that besport can avoid all those forks. :)

I'll do a separate PR if I get more improvements (with idea 2, in particular).

**Comment by Drup** [comment]
@vouillon I would be interested by a comparison of the size of request data for besport too, if it's not too complicated to do.

**Comment by vasilisp** [comment]
Well done @Drup :).

**Pull request merged**